File: backend/app/schemas/batch.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class BatchBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    dance_style: str = Field(default="", max_length=128)
    level: str = Field(default="", max_length=64)
    schedule: str = Field(default="", max_length=255)
    monthly_fee: Decimal = Field(default=Decimal("0.00"), ge=0)
    is_active: bool = True
    notes: str | None = None


class BatchCreate(BatchBase):
    pass


class BatchUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    dance_style: str | None = Field(default=None, max_length=128)
    level: str | None = Field(default=None, max_length=64)
    schedule: str | None = Field(default=None, max_length=255)
    monthly_fee: Decimal | None = Field(default=None, ge=0)
    is_active: bool | None = None
    notes: str | None = None


class BatchOut(BatchBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    student_count: int = 0
    created_at: datetime
    updated_at: datetime


class BatchStats(BatchOut):
    active_student_count: int = 0


class BatchList(BaseModel):
    items: list[BatchOut]
    total: int

File: mobile/lib/models/batch.dart
class Batch {
  final int id;
  final String name;
  final String danceStyle;
  final String level;
  final String schedule;
  final String monthlyFee;
  final bool isActive;
  final String? notes;
  final int studentCount;

  const Batch({
    required this.id,
    required this.name,
    this.danceStyle = '',
    this.level = '',
    this.schedule = '',
    this.monthlyFee = '0',
    this.isActive = true,
    this.notes,
    this.studentCount = 0,
  });

  factory Batch.fromJson(Map<String, dynamic> json) => Batch(
        id: json['id'] as int,
        name: json['name'] as String,
        danceStyle: json['dance_style'] as String? ?? '',
        level: json['level'] as String? ?? '',
        schedule: json['schedule'] as String? ?? '',
        monthlyFee: (json['monthly_fee'] ?? '0').toString(),
        isActive: json['is_active'] as bool? ?? true,
        notes: json['notes'] as String?,
        studentCount: json['student_count'] as int? ?? 0,
      );

  Map<String, dynamic> toJson({bool includeId = false}) => {
        if (includeId) 'id': id,
        'name': name,
        'dance_style': danceStyle,
        'level': level,
        'schedule': schedule,
        'monthly_fee': monthlyFee,
        'is_active': isActive,
        'notes': notes,
      };
}

class BatchListResponse {
  final List<Batch> items;
  final int total;

  const BatchListResponse({required this.items, required this.total});

  factory BatchListResponse.fromJson(Map<String, dynamic> json) => BatchListResponse(
        items: (json['items'] as List)
            .map((e) => Batch.fromJson(e as Map<String, dynamic>))
            .toList(),
        total: json['total'] as int? ?? 0,
      );
}